### PR TITLE
Implementation of encryption and decryption

### PR DESCRIPTION
**Pull request by emmorais**
This PR adds encryption and decryption functionality. It provides basic features to YASHE cryptosystem. It doesn't deal with homomorphic operations.

### CODE REVIEW

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 1]
```diff
@@ -104,14 +118,71 @@ impl<C: PolyConf> Yashe<C> {
...
+                coeff_res.sub_with_borrow(&qm1d2);
+                // TODO: how to divide 2 bigints?
+                coeff_res = coeff_res / q;
+            }
+            // TODO: validate the following
+            // REMARK: since t is a power of 2, we have that coeff_res.0[i] is zero for i > 0
+            // then we can apply the following formula
+            let result = coeff_res.0[0] % self.params.t;
+            res[i] = result.into();
+        }
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:161]: If you implement `DivAssign` as multiplying by `Coeff::inverse()`, then you can replace this entire loop with:
```suggestion
res /= C::t_as_coeff();
```

Just copy these impls, then make the calculation:
```rust
*coeff *= rhs.inverse().expect("can't divide by the zero coefficient");
```

https://github.com/Inversed-Tech/eyelid/blob/3668934d68780513ea61ede8f4ccfb2d6a7eaedb/eyelid-match-ops/src/primitives/poly/modular_poly/mul.rs#L17-L31

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 2]
```diff
@@ -104,14 +118,71 @@ impl<C: PolyConf> Yashe<C> {
...
+    pub fn encrypt(&self, m: Message<C>, public_key: PublicKey<C>, rng: &mut ThreadRng) -> Ciphertext<C> {
+        let logt = self.params.t.ilog2();
+        let s = self.sample_err(rng);
+        let e = self.sample_err(rng);
+        let mut c = s * public_key.h + e;
+        // TODO: find a way to access q and divide by t
+        let mut qdt: BigInt<2> = BigInt::from(Coeff::MODULUS);
+        //let qbt = q / Coeff::from(self.params.t);
+        qdt.divn(logt);
+        let mqdt = m.m * Coeff::from(qdt);
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:130]: Normally you could multiply by `Coeff::inverse()` instead of dividing:
https://docs.rs/ark-ff/0.4.2/ark_ff/fields/trait.Field.html#tymethod.inverse

But that doesn't work for the modulus, because it reduces to zero. Instead, convert to `u128`, and do the arithmetic there.

I also added a `t_as_coeff()` method in PR #79, which could be useful here:
https://github.com/Inversed-Tech/eyelid/blob/3668934d68780513ea61ede8f4ccfb2d6a7eaedb/eyelid-match-ops/src/primitives/yashe/conf.rs#L36-L37

```suggestion
        // We can't divide the modulus using `Coeff` arithmetic, because it reduces to zero.
        let qdt = u128::from(q) / u128::from(t);
        let qdt = C::Coeff::from(qbt);
        let mqdt = m.m * qdt;
```

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 3]
```diff
@@ -104,14 +118,71 @@ impl<C: PolyConf> Yashe<C> {
...
+    }
+
+    /// Sample a polynomial with small random coefficients using a gaussian distribution.
+    #[allow(clippy::cast_possible_truncation)]
+    pub fn sample_err(&self, rng: &mut ThreadRng) -> Poly<C> {
+        self.sample_gaussian(self.params.delta, rng)
+    }
+
+    /// Sample a polynomial with small random coefficients using a gaussian distribution.
+    #[allow(clippy::cast_possible_truncation)]
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:172]: I don't think these lints are needed?

```suggestion
    pub fn sample_err(&self, rng: &mut ThreadRng) -> Poly<C> {
        self.sample_gaussian(self.params.delta, rng)
    }

    /// Sample a polynomial with small random coefficients using a gaussian distribution.
```

**Review by teor2345 — COMMENTED** [review]
Looks good, I think there is a simple way of doing division.

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 2 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe.rs:130]: It might be worth adding `inverse_t()` as a trait method after this method:
https://github.com/Inversed-Tech/eyelid/blob/3668934d68780513ea61ede8f4ccfb2d6a7eaedb/eyelid-match-ops/src/primitives/yashe/conf.rs#L36-L39

With this implementation:
```rust
        // We can't divide the modulus using `Coeff` arithmetic, because it reduces to zero.
        let qdt = u128::from(q) / u128::from(t);
        let qdt = C::Coeff::from(qbt);
```

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 1 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:161]: We need division over the integers, instead of over the finite field. Then `inverse` is not going to help us here, since every element in the field has an inverse. What we need is division with remainder, but we are interested only in the quotient (and not in the remainder).

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 4]
```diff
@@ -104,14 +119,125 @@ impl<C: PolyConf> Yashe<C> {
...
+    pub fn sample_err(&self, rng: &mut ThreadRng) -> Poly<C> {
+        self.sample_gaussian(self.params.delta, rng)
+    }
+
     /// Sample a polynomial with small random coefficients using a gaussian distribution.
     #[allow(clippy::cast_possible_truncation)]
-    pub fn sample_gaussian(&self, rng: &mut ThreadRng) -> Poly<C> {
+    pub fn sample_key(&self, rng: &mut ThreadRng) -> Poly<C> {
+        // standard deviation whose output coefficients are -1, 0, 1 with high probability
+        self.sample_gaussian(self.params.delta/8.0, rng)
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:230]: Let's make `self.params.delta/8.0` into a `KEY_DELTA` constant?

**Comment by teor2345** [comment]
Hey just checking if you need help with this PR?

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 3 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs]: done

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 4 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs]: done

**Comment by emmorais** [comment]
> Hey just checking if you need help with this PR?

Can you please help with with merging?

**Comment by teor2345** [comment]
> > Hey just checking if you need help with this PR?
> 
> Can you please help with with merging?

Yep, just doing this now.

`u128::from<C::Coeff>()` and `u128::into<C::Coeff>()` are already implemented, but we need to require that they are available in the trait type constraints. I'll fix that too.

**Comment by teor2345** [comment]
That was a lot more complicated than I expected, the conversion methods were hidden behind a bunch of traits in different crates. No wonder they were difficult to discover.

I think I've implemented what you wanted, but it was tricky to know without the equations in the function documentation. Feel free to make any fixes or further changes.

File: eyelid-match-ops/src/primitives/poly/modular_poly/mul.rs [hunk 5]
```diff
@@ -19,6 +19,7 @@ impl<C: PolyConf> MulAssign<C::Coeff> for Poly<C> {
         for coeff in &mut self.0.coeffs {
             *coeff *= rhs;
         }
+        self.truncate_to_canonical_form();
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/poly/modular_poly/mul.rs:22]: This fix means we don't need to repeat these checks in the yashe code. (Unless we're working with raw coefficients.)

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 6]
```diff
@@ -104,14 +130,93 @@ where
...
+        let e = self.sample_err(rng);
+        let mut c = s * public_key.h + e;
+
+        // Divide the polynomial coefficient modulus by T, using primitive integer arithmetic.
+        let qdt = C::modulus_as_u128() / C::t_as_u128();
+        let qdt = C::Coeff::from(qdt);
+
+        // Multiply the message by the qdt scalar, and add it to the ciphertext.
+        m.m *= qdt;
+        c += m.m;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:153]: Please check that these are the equations we intend to implement for encryption.

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 7]
```diff
@@ -104,14 +130,93 @@ where
...
+            // Add (Q - 1)/2 to implement rounding rather than truncation
+            coeff_res += C::modulus_minus_one_div_two_as_u128();
+            // Divide by Q
+            coeff_res /= C::modulus_as_u128();
+            // Modulo T
+            coeff_res %= C::t_as_u128();
+
+            // And update the coefficient
+            *coeff = coeff_res.into();
+        }
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:191]: Please check that these are the equations we intend to implement for decryption.

It was tricky to work out what was intended here, so feel free to ask questions or add extra comments.

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 8]
```diff
@@ -104,14 +130,93 @@ where
...
+        Message { m: res }
+    }
+
+    /// Sample a polynomial with small random coefficients using a gaussian distribution.
+    pub fn sample_err(&self, rng: &mut ThreadRng) -> Poly<C> {
+        self.sample_gaussian(C::ERROR_DELTA, rng)
+    }
+
+    /// Sample a polynomial with small random coefficients using a gaussian distribution.
+    /// TODO: this function seems to be returning too few non-zero elements
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:205]: `v as i64` will truncate the `f64` value, discarding any fractional parts.

Did you want to [round halfway](https://doc.rust-lang.org/std/primitive.f64.html#method.round) or [use the next highest integer](https://doc.rust-lang.org/std/primitive.f64.html#method.ceil) before converting?

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 9]
```diff
@@ -121,17 +226,53 @@ where
...
+    where
+        T: SampleUniform,
+        R: SampleRange<T> + Clone,
+        C::Coeff: From<T>,
+    {
+        let mut res = Poly::non_canonical_zeroes(C::MAX_POLY_DEGREE);
+        for i in 0..C::MAX_POLY_DEGREE {
+            let coeff_rand = rng.gen_range(range.clone());
+            res[i] = coeff_rand.into();
+        }
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:273]: We were repeating the same code in two different functions, so it seemed helpful to have a generic sampler for any coefficient range.

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 10]
```diff
@@ -25,27 +29,67 @@ use crate::primitives::poly::modular_poly::conf::TinyTest;
...
+    }
+
+    /// A convenience method to convert [`Coeff::MODULUS_MINUS_ONE_DIV_TWO`](PrimeField::MODULUS_MINUS_ONE_DIV_TWO) to `u128`.
+    fn modulus_minus_one_div_two_as_u128() -> u128 {
+        let modulus: BigUint = Self::Coeff::MODULUS_MINUS_ONE_DIV_TWO.into();
+
+        modulus
+            .to_u128()
+            .expect("constant modulus is small enough for u128")
+    }
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:85]: This conversion was really tricky to find, but it is better to use existing tested methods.

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 11]
```diff
@@ -68,5 +110,8 @@ impl YasheConf for TinyTest {
     const T: u64 = 7;
 
     /// Limited to 1/6 of the modulus, so that the sampled values are valid within 6 sigmas.
-    const DELTA: f64 = 0.9;
+    const KEY_DELTA: f64 = 0.9;
+
+    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
+    const ERROR_DELTA: f64 = 0.3;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:116]: These parameters might need to be adjusted to make the tiny poly tests pass.

**Review by teor2345 — COMMENTED** [review]
Here are the changes I made today in detail.

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 12]
```diff
@@ -25,27 +29,67 @@ use crate::primitives::poly::modular_poly::conf::TinyTest;
 /// Encryption keys and ciphertexts with different parameters are incompatible.
 pub trait YasheConf: PolyConf
```
Comment (naure) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:30]: An alternative: instead of YasheConf being a super-trait of PolyConf, it defines two PolyConf as associated types, for plaintext and ciphertexts.

```rust
pub trait YasheConf {
    type PlainPolyConf: PolyConf;

    type CipherPolyConf: PolyConf;

    …
}

```

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 8 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:205]: round halfway

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 6 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:153]: This is correct.

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 7 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:191]: This is correct.

File: eyelid-match-ops/src/primitives/yashe/conf.rs (hunk 12 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe/conf.rs:30]: PolyConf has 2 items: degree and modulus. But degree must be the same for PlainPolyConf and CipherPolyConf. YasheConf must have a way to configure this common degree, and a way to choose the plaintext and the ciphertext modulus (denoted t and q in the paper). Beyond that, YasheConf has other parameters like standard deviations. It can grow in the future if we start doing things like bootstrapping.

File: eyelid-match-ops/benches/match-ops.rs [hunk 13]
```diff
@@ -429,6 +447,49 @@ pub fn bench_keygen(settings: &mut Criterion) {
...
+    // Setup parameters
+    let mut rng = rand::thread_rng();
+    let ctx: Yashe<TestRes> = Yashe::new();
+
+    let (private_key, public_key) = ctx.keygen(&mut rng);
+    let m = ctx.sample_message(&mut rng);
+    let c = ctx.encrypt(m, public_key, &mut rng);
+
+    settings.bench_with_input(
+        BenchmarkId::new("YASHE dec", "standard parameters with degree N"),
```
Comment (teor2345) [inline eyelid-match-ops/benches/match-ops.rs:482]:
```suggestion
        BenchmarkId::new("YASHE enc", SMALL_RANDOM_NAME),
        &ctx,
        |benchmark, ctx| {
            // To avoid timing dropping the return value, we require it to be returned from the closure.
            benchmark.iter_with_large_drop(|| -> Ciphertext<TestRes> {
                ctx.encrypt(m.clone(), public_key.clone(), &mut rng)
            })
        },
    );
}

/// Run [`Yashe::dec()`] as a Criterion benchmark with random data.
pub fn bench_dec(settings: &mut Criterion) {
    // Setup parameters
    let mut rng = rand::thread_rng();
    let ctx: Yashe<TestRes> = Yashe::new();

    let (private_key, public_key) = ctx.keygen(&mut rng);
    let m = ctx.sample_message(&mut rng);
    let c = ctx.encrypt(m, public_key, &mut rng);

    settings.bench_with_input(
        BenchmarkId::new("YASHE dec", SMALL_RANDOM_NAME),
```

**Comment by emmorais** [comment]
> That was a lot more complicated than I expected, the conversion methods were hidden behind a bunch of traits in different crates. No wonder they were difficult to discover.
> 
> I think I've implemented what you wanted, but it was tricky to know without the equations in the function documentation. Feel free to make any fixes or further changes.

It looks very good now, thanks a lot for improving it in many ways! I think it is ready to merge now.

**Comment by emmorais** [comment]
Feel free to merge after approving.

File: eyelid-match-ops/src/primitives/yashe/conf.rs (hunk 12 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe/conf.rs:30]: Let’s open a ticket for this?
I don’t mind which way we do it, but some ways will be easier in Rust than others.

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 14]
```diff
@@ -62,22 +86,24 @@ where
...
             priv_key.truncate_to_canonical_form();
 
             let priv_key_inv = priv_key.inverse();
 
-            if let Ok(_priv_key_inv) = priv_key_inv {
-                return PrivateKey { f, finv, priv_key };
+            if let Ok(priv_key_inv) = priv_key_inv {
+                return PrivateKey {
+                    f,
+                    priv_key_inv,
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:104]: This field was changed from `f_inv` to `priv_key_inv`, but the test is still checking `f_inv`.

This occasionally causes failures, particularly with tiny polynomials:
https://github.com/Inversed-Tech/eyelid/actions/runs/9072716142/job/24928644893?pr=82#step:5:31
So I’m not sure if we can merge failing tests to the main branch.

Can you update the test so it does what you expect?
I’m not sure what we need to do to test that the keys and encryption are correct, so I don’t want to merge without your input.

File: eyelid-match-ops/src/primitives/yashe/conf.rs (hunk 11 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe/conf.rs:116]: It seems like the test itself is incorrect, see my other comment.

File: eyelid-match-ops/src/primitives/yashe.rs [hunk 15]
```diff
@@ -121,17 +226,53 @@ where
             // This is ok because the C::Coeff modulus is smaller than MIN/MAX.
             res[i] = C::Coeff::from(v as i64);
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe.rs:227]:
```suggestion
            //
            // `as` truncates by default, but we want to round to the nearest integer.
            res[i] = C::Coeff::from(v.round() as i64);
```

**Review by teor2345 — CHANGES_REQUESTED** [review]
I think the occasional failing test might be a blocker for merging.

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 8 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe.rs:205]: Ok, here’s how to do that:
https://github.com/Inversed-Tech/eyelid/pull/82#discussion_r1602322652

(There’s also round ties to even, but ties are low probability, so the difference doesn’t matter.)

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 14 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe.rs:104]: I think we should not test an encryption scheme using tiny parameters. It is not secure when instantiated like that, an even correctness doesn't hold and weird things happen. We should worry making it pass the unit test because a corresponding situation doesn't happen for real parameters. Can we stop using tiny parameters for YASHE tests?

File: eyelid-match-ops/src/primitives/yashe/conf.rs (hunk 12 above)
Comment (emmorais) [reply eyelid-match-ops/src/primitives/yashe/conf.rs:30]: [Here](https://github.com/Inversed-Tech/eyelid/issues/90) it is.

File: eyelid-match-ops/src/primitives/yashe.rs (hunk 14 above)
Comment (teor2345) [reply eyelid-match-ops/src/primitives/yashe.rs:104]: This is now ticket #91.

**Review by teor2345 — APPROVED** [review]
Looks good after the test fixes!

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 16]
```diff
@@ -25,27 +29,67 @@ use crate::primitives::poly::modular_poly::conf::TinyTest;
...
 where
-    Self::Coeff: From<u64> + From<i64>,
+    // The `Field` trait is already `From<u128> + From<u64>` (and all the other unsigned types).
+    // The `Fp` types are `From<i64>` (and all the other signed types).
+    // But there are no trait bounds guaranteeing these conversions, so we need to require them.
+    // Unfortunately, these bounds also need to be copied to each generic type and impl block.
+    Self::Coeff: From<u128> + From<u64> + From<i64>,
 {
     /// The plaintext coefficient modulus
     const T: u64;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:39]:
```suggestion
    /// The plaintext coefficient modulus.
    /// Must be a power of two, and smaller than the modulus.
    const T: u64;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 17]
```diff
@@ -25,27 +29,67 @@ use crate::primitives::poly::modular_poly::conf::TinyTest;
...
+            .to_u128()
+            .expect("constant modulus is small enough for u128")
+    }
 }
 
 /// Iris bit length polynomial parameters.
 ///
 /// This uses the full number of iris bits, which gives an upper bound on benchmarks.
 impl YasheConf for IrisBits {
     const T: u64 = IRIS_BIT_LENGTH as u64;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:92]:
```suggestion
    const T: u64 = 2048;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 18]
```diff
@@ -68,5 +110,8 @@ impl YasheConf for TinyTest {
     const T: u64 = 7;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:110]:
```suggestion
    const T: u64 = 4;
```

**Review by teor2345 — COMMENTED** [review]
Changes to use correct T

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 19]
```diff
@@ -6,6 +6,10 @@
 //! RUSTFLAGS="--cfg tiny_poly" cargo bench --features benchmark
 //! ```
 
+use ark_ff::PrimeField;
+use num_bigint::BigUint;
+use num_traits::ToPrimitive;
+
 use crate::{
     primitives::poly::{modular_poly::conf::IrisBits, PolyConf},
     IRIS_BIT_LENGTH,
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:15]:
```suggestion
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 20]
```diff
@@ -6,9 +6,12 @@
...
 //! ```
 
+use ark_ff::PrimeField;
+use num_bigint::BigUint;
+use num_traits::ToPrimitive;
+
 use crate::{
     primitives::poly::{modular_poly::conf::IrisBits, PolyConf},
-    IRIS_BIT_LENGTH,
 };
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:16]:
```suggestion

```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 21]
```diff
@@ -6,9 +6,12 @@
 //! RUSTFLAGS="--cfg tiny_poly" cargo bench --features benchmark
 //! ```
 
+use ark_ff::PrimeField;
+use num_bigint::BigUint;
+use num_traits::ToPrimitive;
+
 use crate::{
     primitives::poly::{modular_poly::conf::IrisBits, PolyConf},
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:14]:
```suggestion
use crate::primitives::poly::{modular_poly::conf::IrisBits, PolyConf};
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 22]
```diff
@@ -6,11 +6,11 @@
...
-use crate::{
-    primitives::poly::{modular_poly::conf::IrisBits, PolyConf},
-    IRIS_BIT_LENGTH,
-};
+use ark_ff::PrimeField;
+use num_bigint::BigUint;
+use num_traits::ToPrimitive;
 
+use crate::primitives::poly::{modular_poly::conf::IrisBits, PolyConf};
 pub use crate::primitives::poly::modular_poly::conf::TestRes;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:14]:
```suggestion
use crate::primitives::poly::{modular_poly::conf::IrisBits, PolyConf};
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 23]
```diff
@@ -6,11 +6,11 @@
...
 
 #[cfg(not(tiny_poly))]
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:16]:
```suggestion
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 24]
```diff
@@ -54,8 +95,6 @@ impl YasheConf for IrisBits {
...
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:95]:
```suggestion
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 25]
```diff
@@ -65,8 +101,11 @@ impl YasheConf for FullRes {
 #[cfg(tiny_poly)]
 impl YasheConf for TinyTest {
     /// Limited to the modulus of the underlying `Coeff` type.
-    const T: u64 = 7;
+    const T: u64 = 4;
 
     /// Limited to 1/6 of the modulus, so that the sampled values are valid within 6 sigmas.
-    const DELTA: f64 = 0.9;
+    const KEY_DELTA: f64 = 0.9;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:107]:
```suggestion
    const KEY_DELTA: f64 = 0.6;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 26]
```diff
@@ -65,8 +101,11 @@ impl YasheConf for FullRes {
...
+
+    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
+    const ERROR_DELTA: f64 = 0.3;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:110]:
```suggestion
    const ERROR_DELTA: f64 = 0.2;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 27]
```diff
@@ -65,8 +101,11 @@ impl YasheConf for FullRes {
...
     /// Limited to the modulus of the underlying `Coeff` type.
-    const T: u64 = 7;
+    const T: u64 = 4;
 
     /// Limited to 1/6 of the modulus, so that the sampled values are valid within 6 sigmas.
-    const DELTA: f64 = 0.9;
+    const KEY_DELTA: f64 = 0.6;
+
+    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
+    const ERROR_DELTA: f64 = 0.2;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:110]:
```suggestion
    const KEY_DELTA: f64 = 0.5;

    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
    const ERROR_DELTA: f64 = 0.16;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 28]
```diff
@@ -65,8 +101,11 @@ impl YasheConf for FullRes {
...
     /// Limited to the modulus of the underlying `Coeff` type.
-    const T: u64 = 7;
+    const T: u64 = 4;
 
     /// Limited to 1/6 of the modulus, so that the sampled values are valid within 6 sigmas.
-    const DELTA: f64 = 0.9;
+    const KEY_DELTA: f64 = 0.5;
+
+    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
+    const ERROR_DELTA: f64 = 0.16;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:110]:
```suggestion
    const KEY_DELTA: f64 = 0.3;

    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
    /// This effectively makes the error zero.
    const ERROR_DELTA: f64 = 0.1;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 29]
```diff
@@ -65,8 +101,12 @@ impl YasheConf for FullRes {
...
-    const T: u64 = 7;
+    const T: u64 = 4;
 
     /// Limited to 1/6 of the modulus, so that the sampled values are valid within 6 sigmas.
-    const DELTA: f64 = 0.9;
+    const KEY_DELTA: f64 = 0.3;
+
+    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
+    /// This effectively makes the error zero.
+    const ERROR_DELTA: f64 = 0.1;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:111]:
```suggestion
    /// This effectively sets the key randomness to zero, and the key to one, to six sigma probability.
    const KEY_DELTA: f64 = 0.01;

    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
    /// This effectively makes the error zero, to six sigma probability.
    const ERROR_DELTA: f64 = 0.01;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 30]
```diff
@@ -25,48 +23,89 @@ use crate::primitives::poly::modular_poly::conf::TinyTest;
...
 /// Tiny test polynomials, used for finding edge cases in tests.
 ///
 /// The test parameters are specifically chosen to make failing tests easy to read and diagnose.
-
+/// TODO: these parameters don't work for encryption and decryption, find some that do.
 #[cfg(tiny_poly)]
 impl YasheConf for TinyTest {
     /// Limited to the modulus of the underlying `Coeff` type.
-    const T: u64 = 7;
+    const T: u64 = 6;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:104]:
```suggestion
    const T: u64 = 4;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 31]
```diff
@@ -25,48 +23,89 @@ use crate::primitives::poly::modular_poly::conf::TinyTest;
...
 
     /// Limited to 1/6 of the modulus, so that the sampled values are valid within 6 sigmas.
-    const DELTA: f64 = 0.9;
+    const KEY_DELTA: f64 = 0.9;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:107]:
```suggestion
    const KEY_DELTA: f64 = 0.6;
```

File: eyelid-match-ops/src/primitives/yashe/conf.rs [hunk 32]
```diff
@@ -25,48 +23,89 @@ use crate::primitives::poly::modular_poly::conf::TinyTest;
...
+
+    /// Limited to 1/3 of KEY_DELTA, so that the error is small enough for valid decryption.
+    const ERROR_DELTA: f64 = 0.3;
```
Comment (teor2345) [inline eyelid-match-ops/src/primitives/yashe/conf.rs:110]:
```suggestion
    const ERROR_DELTA: f64 = 0.2;
```
